winml/lib/Api/LearningModelSession.cpp

```diff
@@ -177,74 +172,72 @@ auto CreateBinding(
-}
-
-uint64_t
-LearningModelSession::Run(
-    winrt::com_ptr<winmlp::LearningModelBinding> binding_impl) {
+uint64_t LearningModelSession::Run(winrt::com_ptr<winmlp::LearningModelBinding> binding_impl) {
   CheckClosed();
+
   auto device = device_.as<LearningModelDevice>();
   CWinMLAutoLock lock(!device->IsCpuDevice() ? &evaluate_lock_ : nullptr);
-  // TODO : set the run_options
-  Ort::RunOptions run_options;
+
   binding_impl->BindUnboundOutputs();
 
-  std::vector<const char*> inputNames_c;
-  for (int i=0; i < binding_impl->GetInputNames().size(); i++)
-  {
-    inputNames_c.push_back(binding_impl->GetInputNames()[i].c_str());
-  }
-  std::vector<const char*> outputNames_c;
-  for (int i = 0; i < binding_impl->GetOutputNames().size(); i++) {
-    outputNames_c.push_back(binding_impl->GetOutputNames()[i].c_str());
-  }
-  OrtSession* session = nullptr;
-
-  WINML_THROW_IF_FAILED(inference_session_->GetOrtSession(&session));
-  // Invoke run on the ORT session.
-  Ort::ThrowOnError(OrtRun(
-    session, 
-    run_options, 
-    inputNames_c.data(),
-    binding_impl->GetInputs().data(),
-    binding_impl->GetInputs().size(),
-    outputNames_c.data(),
-    binding_impl->GetOutputs().data(),
-    binding_impl->GetOutputs().size()));
+  auto& input_names = binding_impl->GetInputNames();
+  std::vector<const char*> input_names_raw;
+  std::transform(
+      std::begin(input_names),
+      std::end(input_names),
+      std::back_inserter(input_names_raw),
+      [&](auto& name) { return name.c_str(); });
+
+  auto& inputs = binding_impl->GetInputs();
+  std::vector<WinML::IValue*> inputs_raw;
+  std::transform(
+      std::begin(inputs),
+      std::end(inputs),
+      std::back_inserter(inputs_raw),
+      [&](auto& input) { return input.get(); });
+
+  auto& output_names = binding_impl->GetOutputNames();
+  std::vector<const char*> output_names_raw;
+  std::transform(
+      std::begin(output_names),
+      std::end(output_names),
+      std::back_inserter(output_names_raw),
+      [&](auto& name) { return name.c_str(); });
+
+  auto outputs = binding_impl->GetOutputs();
+  std::vector<WinML::IValue*> outputs_raw;
+  std::transform(
+      std::begin(outputs),
+      std::end(outputs),
+      std::back_inserter(outputs_raw),
+      [&](auto& input) { return input.get(); });
+
+  engine_->Run(input_names_raw.data(),
+               inputs_raw.data(),
+               input_names_raw.size(),
+               output_names_raw.data(),
+               outputs_raw.data(),
+               output_names_raw.size());
 
   if (!device->IsCpuDevice()) {
     // Flush the D3D12 work from the DML execution provider and queue a fence before we release the lock.
     // This allows us to wait without holding onto the lock in GetResults.
-    inference_session_->FlushContext(GetExecutionProvider());
+    engine_->FlushContext();
```

Comment:
>FlushContext [](start = 13, length = 12)

This check can probably be moved into the the ort lib

Reply:
For a future refactor...

---
In reply to: [370319154](https://github.com/microsoft/onnxruntime/pull/2891#discussion_r370319154) [](ancestors = 370319154)